Fix the memory growth

da-kami opened the pull request:
This is an attempt to fix the memory growth caused by `rust-electrum-client`'s `batch_script_get_history` implementation. We chunk up the scripts to be checked into batches and call `script_get_history` for the batches in parallel in separate worker threads.

The results are communicated back using a channel, and are processed as they come; we don't have to rely on any order here.

---

The implementation should be complete and compiles, but there are still a few rough edges ;)
@klochowicz please run this with profiling so we can see if it has the expected behavior.

TODO:

- [x] Fix tests: The tests have to be adapted to cover the new behavior. 
- [x] Evaluate if the buffer size for the channel should be in relation to the batch size

da-kami committed:
Fix the memory growth by self-implemented batching

This is an attempt to fix the memory growth caused by `rust-electrum-client`'s `batch_script_get_history` implementation.
We chunk up the scripts to be checked into batches and call `script_get_history` for the batches in parallel in separate worker threads.

The results are communicated back using a channel, and are processed as they come; we don't have to rely on any order here.

bonomat commented on crates/daemon/src/monitor.rs:
```diff
@@ -466,25 +470,76 @@ impl Actor {
+                let sender = sender.clone();
+                move || {
+                    for script in batch {
+                        // TODO: Evaluate if we should notify the receivers about errors, I feel
```
Agree, as long as we don't intend to deal with the errors there is no need in bubbling it up

klochowicz replied:
there are some errors that we might be interested in, e.g. if we're exceeding the rate, we could slow down the requests.

da-kami replied:
> there are some errors that we might be interested in, e.g. if we're exceeding the rate, we could slow down the requests.

Is it fair to say that we can add such a feature later, once we know how this performs on testnet? 
I find it hard to design this without knowing how the solution will perform under load.

We might also want to change the `BATCH_SIZE` - 25 is a somewhat arbitrary number. At the moment we process batches of 25 requests in parallel, for `7000` subscriptions this means `280` worker threads in parallel. When running the test that simulates this behavior this did not cause problems on our local machines, but there was CPU load. We will have to see how this performs on the testnet machine.

bonomat commented on crates/daemon/src/monitor.rs:
```diff
@@ -466,25 +470,76 @@ impl Actor {
-        );
+        // TODO: Evaluate buffer; if we do batches of 25 and allow buffer of 100 here, does that
+        // mean we can at most process 4 batches at the time?  What happens if the buffer is
+        // exhausted? (I would expect the next messages will be queued) We use tokio::sync::
```
From the docs: 

> The channel will buffer up to the provided number of messages. Once the buffer is full, attempts to send new messages will wait until a message is received from the channel. The provided buffer capacity must be at least 1.

so you are right with your assumption.

bonomat commented on crates/daemon/src/monitor.rs:
```diff
@@ -466,25 +470,76 @@ impl Actor {
+
+        let client = self.client.clone();
+
+        let batch_futures = batches.into_iter().map(move |batch| {
```
🔧
I think you will need to wait between the batches, otherwise you just fire one after the other because you immediately spawn them away.

klochowicz replied:
It would be interesting to see whether we actually need to wait here; we're not concerned about the results coming out of order, as long as the results will come back (depends on the behaviour of Electrum?)

bonomat replied:
If we don't wait here, we don't have to use batches because they get all spawned away. Or am I misreading this?

klochowicz replied:
well, the difference is still that we're partitioning them in different worker threads (the `spawn_blocking`)

bonomat replied:
So we will spawn batches of 25 in parallel and execute all (4x) in parallel and wait until at least 1 is finished before we spawn another 25?

klochowicz replied:
pretty much, but the 4x figure depends on the limit of the worker threads. AFAICT, this limit is by default 512, so we pretty much fire all the batches at the same time, and then reach the electrum API one at a time from all these threads.
tokio has separate threads, separate from the runtime that are used here.

bonomat replied:
gotcha, thanks for the explanation. :)

da-kami replied:
Maybe I misunderstand what you are saying.

The idea of this implementation is to chunk up the requests in packages of 25 and then process all these packages (I called them batches) in parallel.
Maybe the term `batch` is not used correctly here, because we are not processing what is in the batch in parallel, but we process all the chunks in parallel. 
I am happy to rename what is called `batch` at the moment. 
Do you think the implementation needs changes?

---

Note that in fixup https://github.com/itchysats/itchysats/pull/3190/commits/c009237fd04b7258b066e4d867323d77fd5d6785 we stop joining all the `spawn_blocking` join handles, because that is not necessary and just adds complexity.

bonomat reviewed (commented):
We should also change how often we call `Sync`.

da-kami committed:
fixup! Fix the memory growth by self-implemented batching

Separate the logic for processing events (i.e. CFD state updates) from the sync.
Additionally, resolve some other TODOs.

da-kami committed:
Explicitly set electrum client timeout

da-kami committed:
fixup! Fix the memory growth by self-implemented batching

Make the tests pass.

klochowicz commented on crates/daemon/src/monitor.rs:
```diff
@@ -466,66 +485,46 @@ impl Actor {
+                                    tx_script_updates.blocking_send(script_history_response)
+                                {
+                                    tracing::error!(
+                                        "Error when processing script_get_history response: {e:#}"
```
🛠 this one gets fired off all the time when I tested it with response "channel closed", and we never stop.

I don't think just logging the errors here is an option....

da-kami replied:
I was unable to directly reproduce the "channel closed" issue, but I found a problem in the implementation where the sender and the receiver were not properly tied together, because the sender was not moved correctly. 
I think that this issue caused this behavior. In the PRs replacing this one I also added tests to ensure that the underlying batch functionality works, but we yet have to test the complete production code again. 

I also thought about adding a test that spins up the complete monitor Actor, but that is a bit too much for now...

bonomat commented on crates/sqlite-db/src/lib.rs:
```diff
@@ -706,6 +706,8 @@ async fn load_cfd_events(
                 rollover::load(&mut *conn, *cfd_row_id, *event_row_id).await?
             {
+                let dlc_id = dlc.commit.0.txid();
+                tracing::warn!("Attaching DLC {dlc_id} from event {event_row_id} to CFD {id}");
```
Why is this a `warn`?

da-kami replied:
This is a weird rebasing mistake, not sure how that sneaked in but it is removed in the PRs that replace this one :)
